src/RecurringGrantDropWithRegistrar.sol

```diff
@@ -0,0 +1,193 @@
+  /// @param _grantRegistrar The GrantRegistrar that will gate the claims
+  /// @param _token The ERC20 token that will be airdropped
+  /// @param _holder The address holding the tokens that will be airdropped
+  /// @param _grant The grant that contains the amounts and validity
+  event RecurringGrantDropInitialized(
+    GrantRegistrar _grantRegistrar,
+    ERC20 _token,
+    address _holder,
+    IGrant _grant
+  );
+
+  /// @notice Emitted when a grant is successfully claimed
+  /// @param receiver The address that received the tokens
+  event GrantClaimed(uint256 grantId, address receiver);
+
+  /// @notice Emitted when the grantRegistrar is changed
+  /// @param grantRegistrar The new GrantRegistrar instance
+  event GrantRegistrarUpdated(GrantRegistrar grantRegistrar);
+
+  /// @notice Emitted when the token is changed
+  /// @param token The new token
+  event TokenUpdated(ERC20 token);
+
+  /// @notice Emitted when the holder is changed
+  /// @param holder The new holder
+  event HolderUpdated(address holder);
+
+  /// @notice Emitted when the grant is changed
+  /// @param grant The new grant instance
+  event GrantUpdated(IGrant grant);
+
+  ///////////////////////////////////////////////////////////////////////////////
+  ///                               CONSTRUCTOR                              ///
+  //////////////////////////////////////////////////////////////////////////////
+
+  /// @notice Deploys a WorldIDAirdrop instance
+  /// @param _grantRegistrar The GrantRegistrar that will gate the claims
+  /// @param _token The ERC20 token that will be airdropped
+  /// @param _holder The address holding the tokens that will be airdropped
+  /// @param _grant The grant that contains the amounts and validity
+  constructor(
+    GrantRegistrar _grantRegistrar,
+    ERC20 _token,
+    address _holder,
+    IGrant _grant
+  ) Ownable(msg.sender) {
+    if (address(_grantRegistrar) == address(0)) revert InvalidConfiguration();
+    if (address(_token) == address(0)) revert InvalidConfiguration();
+    if (address(_holder) == address(0)) revert InvalidConfiguration();
+    if (address(_grant) == address(0)) revert InvalidConfiguration();
+
+    grantRegistrar = _grantRegistrar;
+    token = _token;
+    holder = _holder;
+    grant = _grant;
+
+    emit RecurringGrantDropInitialized(grantRegistrar, token, holder, grant);
+  }
+
+  ///////////////////////////////////////////////////////////////////////////////
+  ///                               CLAIM LOGIC                               ///
+  //////////////////////////////////////////////////////////////////////////////
+
+  /// @notice Claim the airdrop
+  /// @param grantId The grant ID to claim
+  /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+  function claim(uint256 grantId, address receiver) external {
+    checkClaim(grantId, receiver);
+
+    hasClaimedGrant[grantId][receiver] = true;
```

Comment:
Would suggest reusing storage slots instead, this will be the "hot path" we want to save as much gas on as possible. If we don't do reservations with this contract (i.e. going back in time), we just need to store the current grantId and then check if it's smaller than the current in `checkClaim()`. For reservations we could do bit packing.